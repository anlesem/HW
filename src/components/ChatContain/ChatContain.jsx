import { useEffect, useRef, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { getChatList, getChatCounter } from '../../store/chats/selectors';
import {
  addChatThunk,
  renameChatThunk,
  deleteChatThunk,
  initChatsDataThunk
} from '../../store/chats/actions';
import { initMessageList, initTempInput } from '../../store/messages/actions';

import style from './ChatContain.module.scss';

import { ChatList } from '../ChatList/ChatList';
import { ChatButtons } from '../ChatButtons/ChatButtons';
import { ChatForm } from '../ChatForm/ChatForm';

export const ChatContain = () => {
  const dispatch = useDispatch();
  const chats = useSelector(getChatList);
  const counter = useSelector(getChatCounter);
  const [checked, setChecked] = useState([0]);
  const scrollChats = useRef(null);

  const handleToggle = (value) => () => {
    const currentIndex = checked.indexOf(value);
    const newChecked = [...checked];
    if (currentIndex === -1) {
      newChecked.push(value);
    } else {
      newChecked.splice(currentIndex, 1);
    }
    setChecked(newChecked);
  };

  const handleAdd = () => {
    dispatch(addChatThunk(counter + 1));
    dispatch(initMessageList(counter + 1));
    dispatch(initTempInput);
    scrollChats.current.scrollTop =
      scrollChats.current.scrollHeight - scrollChats.current.clientHeight;
  };

  const handleDelete = (checked) => {
    checked.forEach((id) => {
      dispatch(deleteChatThunk(id));
    });
    setChecked([0]);
  };

  const changeNameChat = (text) => {
    dispatch(renameChatThunk(checked[1], text));
    setChecked([0]);
  };

  useEffect(() => {
    dispatch(initChatsDataThunk());
  }, []);

  return (
    <div className={style.wrap}>
      <ChatList data={{ handleToggle, chats, checked, scrollChats }} />
      <ChatButtons data={{ handleAdd, handleDelete, checked }} />
      {checked.length === 2 && <ChatForm data={{ changeNameChat, checked }} />}
    </div>
  );
};
