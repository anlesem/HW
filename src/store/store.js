import { createStore, combineReducers, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import { profileReducer } from './profile/reducer';
import { chatsReducer } from './chats/reducer';
import { messagesReducer } from './messages/reducer';
import { dataReducer } from './data/reducer';

const persistConfig = {
  key: 'root',
  storage,
  blacklist: ['data']
};

const rootReducers = combineReducers({
  profile: profileReducer,
  chats: chatsReducer,
  messages: messagesReducer,
  data: dataReducer
});

const persistedReducers = persistReducer(persistConfig, rootReducers);

export const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const store = createStore(persistedReducers, composeEnhancers(applyMiddleware(thunk)));

export const persistor = persistStore(store);
