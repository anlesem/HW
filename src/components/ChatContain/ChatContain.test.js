import { BrowserRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import { ChatContain } from './ChatContain';

import { Provider } from 'react-redux';
import { store } from '../../store/store';

describe('ChatContain', () => {
  it('Компонент существует', () => {
    expect(ChatContain).toBeInstanceOf(Function);
  });

  it('Снимок состояния', () => {
    const chatId = 1;

    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <ChatContain chatData={chatId} />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment(<ChatContain />)).toMatchSnapshot();
  });

  it('Список чатов', () => {
    const chatId = 1;
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ChatContain chatData={chatId} />
        </BrowserRouter>
      </Provider>
    );

    expect(screen.getByRole('list')).toBeTruthy();
    expect(screen.getAllByRole('listitem')).toHaveLength(1);
  });
});
