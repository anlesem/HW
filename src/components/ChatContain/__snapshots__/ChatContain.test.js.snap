// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatContain Снимок состояния 1`] = `
<DocumentFragment>
  <div
    class="wrap"
  >
    <ul
      class="MuiList-root MuiList-padding chatList css-1mcnwpj-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters css-g28sfz-MuiListItem-root"
      >
        <div
          class="MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiButtonBase-root css-1m4u3pv-MuiButtonBase-root-MuiListItemButton-root"
          data-testid="chat-item-1"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <span
              class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root PrivateSwitchBase-edgeStart css-zkqepn-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                aria-labelledby="checkbox-list-label-1"
                class="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate="false"
                data-testid="chat-check-1"
                tabindex="-1"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
            id="checkbox-list-label-1"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
            >
              чат 1
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
        <div
          class="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
        >
          <a
            class=""
            href="/chats/1"
          >
            <button
              aria-label="comments"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              data-testid="chat-change-1"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CommentIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M21.99 4c0-1.1-.89-2-1.99-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14l4 4-.01-18zM18 14H6v-2h12v2zm0-3H6V9h12v2zm0-3H6V6h12v2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
        </div>
      </li>
    </ul>
    <div
      class="manage css-e53awj-MuiStack-root"
    >
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root Mui-disabled  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        data-testid="chats-button-remove"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DeleteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </span>
        Удалить
      </button>
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        data-testid="chats-button-add"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        Добавить
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
